src/poly/cse.rs

```diff
@@ -1,134 +1,37 @@
-use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
-
 use crate::field::Field;
 
-use std::{
-    collections::HashMap,
-    fmt::Debug,
-    hash::Hash,
-    rc::{Rc, Weak},
-};
-
-use super::{Expr, HashResult, VarAssignments};
-
-/// Common Subexpression Elimination - takes a collection of expr
-pub fn cse<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    exprs: Vec<Expr<F, V, ()>>,
-    queriables: &Vec<V>,
-) -> Vec<Rc<Expr<F, V, HashResult>>> {
-    // generate random point in the field set
-    let assignments = generate_random_assignment(queriables);
-
-    // hash table with the hash of the expression as the key
-    // and the weak reference to the expression as the value
-    let mut seen_hashes: HashMap<u64, Weak<Expr<F, V, HashResult>>> = HashMap::new();
-
-    let mut result = Vec::new();
-
-    for expr in exprs {
-        let hashed_expr = Rc::new(expr.hash(&assignments));
-        let hash = hashed_expr.meta().hash;
-
-        // if the hash is already in the hash table,
-        // push the existing expression to the result
-        if let Some(existing_weak) = seen_hashes.get(&hash) {
-            if let Some(existing_expr) = existing_weak.upgrade() {
-                result.push(existing_expr);
-                continue;
-            }
-        }
-
-        // otherwise, insert the hash and the weak reference to the expression
-        seen_hashes.insert(hash, Rc::downgrade(&hashed_expr));
-        result.push(hashed_expr);
-    }
-
-    result
-}
-
-fn generate_random_assignment<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    queriables: &Vec<V>,
-) -> VarAssignments<F, V> {
-    let mut rng = ChaCha20Rng::seed_from_u64(0);
-
-    let mut assignments = HashMap::new();
-
-    for queriable in queriables {
-        let value = F::random(&mut rng);
-        assignments.insert(queriable.clone(), value);
-    }
-
-    assignments
+use super::{ConstrDecomp, Expr, HashResult, SignalFactory};
+use std::{fmt::Debug, hash::Hash};
+
+pub fn replace_expr<F: Field + Hash, V: Clone + Eq + Hash + Debug, SF: SignalFactory<V>>(
+    expr: &Expr<F, V, HashResult>,
+    common_exprs: &[Expr<F, V, HashResult>],
```

Comment:
@rutefig please put common_expr only, so we can first substitue the higher degree subexpressions. We also after substituting have to see if there are common sub expression that are not there anymore because there were part of the substituted subexpression

Reply:
so what is happening actually is on the `compiler/cse` mod I'm identifying all of the common subexpressions for that step based on the minimum degree and minimum of occurrences (which we can define) and this method is doing what you're saying, so, we are replacing the `expr`, but I'm passing a reference to the identified common subexpressions just to check for that  `expr` if it's a common se or not, and if it's not I start to check for each subexpression recursively on that expr. Does it make sense? Because I feel what you're saying is what is here but maybe it's not the clearest way. So I can try to improve that if you want, but what is happening is exactly what you mentioned

Reply:
Yes, I read the code. We should not pass all the common sub expresions, only one, and from outside decide which one we start substituting. And then we have to check the situation again.